Build OCI image to use CLI in containers

aj-stein opened the pull request:
# Committer Notes

Closes #22.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/metaschema-framework/oscal-cli/blob/main/CONTRIBUTING.md) guidance?
- [x] Have you followed the guidelines in our [Contributing](https://github.com/metaschema-framework/oscal-cli/blob/main/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/metaschema-framework/oscal-cli/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- ~Have you written new tests for your core changes, as applicable?~
- ~Have you included examples of how to use your new feature(s)?~
- ~Have you updated all website and readme documentation affected by the changes you made?~


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Builds and publishes multi-architecture container images with optional registry push and attestations; images verify the CLI at startup.

* **Chores**
  * CI archives build artifacts and adds a dedicated container build job; adds scheduled container image cleanup and Dependabot updates for container tooling.
  * Adds multi-stage Docker build configurations for CI and local development.

* **Documentation**
  * README updated with instructions for pulling and running the pre-built CLI container.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aj-stein committed:
[ci] Build official OCI image in GitHub Actions

aj-stein commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+    </configuration>
+    </toolchain>
+</toolchains>
+M2TEMPLATE
```
Good feedback, fixed in https://github.com/metaschema-framework/oscal-cli/pull/208/commits/91cd7b49b0d3855fde185a34fd2f4e6b462ebbb0.

aj-stein commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+COPY pom.xml ./target/*.zi[p] /tmp/
+WORKDIR /tmp
+RUN unzip *.zip -d /opt/oscal-cli-extended
+
```
Good rec, fixed in https://github.com/metaschema-framework/oscal-cli/pull/208/commits/a16e76ac17d952ec3ebf7c5276d299bff82c0618.

aj-stein commented:
Sigh, I will need to go back to the drawing board with this approach, it is more brittle than I thought. I will rubber duck debug with Claude and redesign it to be less brittle.

aj-stein committed:
[ci] Add OCI syntax headers per @coderabbitai rec

aj-stein committed:
[ci] Try multi-platform buildx setup, per @coderabbitai rec

aj-stein commented on Dockerfile.local:
```diff
@@ -0,0 +1,40 @@
+    </configuration>
+    </toolchain>
+</toolchains>
+M2TEMPLATE
```
Resolved per https://github.com/metaschema-framework/oscal-cli/pull/208/commits/4c57d430da21a3cc6e03a07b26e3a6416fd3b03d.

aj-stein commented on .github/workflows/build.yml:
```diff
@@ -105,6 +116,62 @@ jobs:
+        labels: ${{ steps.image_metadata.outputs.annotations }}
+        platforms: linux/amd64
+        cache-from: type=gha
+        cache-to: type=gha,mode=max
```
I did forget this from the original PR feedback from Dave, thanks for the rec.

aj-stein replied:
Resolved in https://github.com/metaschema-framework/oscal-cli/pull/208/commits/317f9a2b0860efbd6cdfa223022d120191de89c1.

david-waltermire replied:
Addressed in commit d04cb42. Added fork PR check to skip container builds for fork PRs (no write access to org container registry). The push is controlled via the \ parameter in the reusable workflow, matching metaschema-java's approach.

aj-stein committed:
[ci] Add COMMIT_MAVEN guardrails to registry push and attestation steps, per @coderabbitai rec

aj-stein commented:
It's only been like a year @david-waltermire, but I did it since we're getting the band back together so to speak. Thanks for waiting for such a speedy PR on my part. 😆

aj-stein committed:
Change approach with separate workflows, per @david-waltermire's feedback

aj-stein committed:
Add explicit permissions for OCI builder job

aj-stein committed:
Remove guard condition for branch checking

aj-stein committed:
Upload build zip to download in container workflow

aj-stein committed:
Now download in container step, don't check out source

aj-stein committed:
Woops, of course you need source for Dockerfile, so add it back

aj-stein committed:
Be sure to align image platforms, per @coderabbitai feedback

aj-stein committed:
Add explicit path for extract build_zip artifact

aj-stein committed:
Follow Copilot feedback, no alpine images for multi-platform builds

aj-stein committed:
Forgot to add unzip install to ci Dockerfile after refactoring split

aj-stein committed:
We switched from Alpine to default (Ubuntu), remember?

aj-stein committed:
Change job names for aesthetics

aj-stein committed:
Heed @coderabbit review, no root user in container

aj-stein committed:
Fix heredoc for local execution

aj-stein committed:
Only make builds from main get latest tag

aj-stein committed:
Follow docker convention, make /app default workdir for mounts

aj-stein committed:
Take @coderabbitai rec, properly order uses before with for GHA workflow schema

aj-stein committed:
Push attestations to registry also by default

aj-stein committed:
Add install instructions for using container images

david-waltermire committed:
Sync container approach with metaschema-java

- Update Docker action versions to latest
- Add chown for /app directory in Dockerfiles
- Add container-cleanup.yml workflow for image pruning
- Add docker ecosystem to dependabot.yml
- Skip container build for fork PRs (no registry access)

david-waltermire committed:
fix: remove unnecessary -it flags from docker run example

The -it flags allocate an interactive terminal which is unnecessary
for the --version command that executes and exits immediately.

david-waltermire commented on README.md:
````diff
@@ -70,6 +70,16 @@ echo 'export PATH=$PATH:/opt/oscal-cli/bin' >> ~/.bashrc
+
+```sh
+docker pull ghcr.io/metaschema-framework/oscal-cli:latest
+docker run -it ghcr.io/metaschema-framework/oscal-cli:latest --version
````
@coderabbitai Addressed in commit 0421b0a. Removed unnecessary -it flags from the docker run example.

aj-stein reviewed (commented):
I can't directly approve as an author, @david-waltermire. It looks good to me, please merge.

I was unable to quickly see some context on mobile when I looked earlier. I look forward to using the container work, so 🚀 🚢 !

aj-stein commented on .github/workflows/container-cleanup.yml:
```diff
@@ -0,0 +1,30 @@
+    steps:
+    - name: Delete old container images
+      uses: snok/container-retention-policy@3b0972b2276b171b212f8c4efbca59ebba26eceb
+      with:
+        account: metaschema-framework
+        image-names: oscal-cli
+        cut-off: 7d
+        keep-n-most-recent: 5
+        image-tags: "!latest !v*"
+        tag-selection: both
+        dry-run: ${{ inputs.dry-run || false }}
+        token: ${{ secrets.GITHUB_TOKEN }}
```
I see that permission, so this must be outdated, @coderabbitai.

david-waltermire replied:
@coderabbitai This uses the same approach as metaschema-java's container-cleanup.yml. GITHUB_TOKEN with packages:write permission should work for org-owned packages when the workflow runs from the same org. If it doesn't work in practice, a PAT can be added later.
